### PR TITLE
SLATE package: make MPI and OpenMP a requirement

### PR DESCRIPTION
**Pull request by G-Ragghianti**
In response to issue https://github.com/spack/spack/issues/44952 .

### CODE REVIEW

File: var/spack/repos/builtin/packages/slate/package.py [hunk 1]
```diff
@@ -101,6 +101,7 @@ class Slate(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("rocsolver", when="+rocm")
 
     requires("%oneapi", when="+sycl", msg="slate+sycl must be compiled with %oneapi")
+    requires("+mpi", msg="MPI is required")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/slate/package.py]: Hmm, this makes little sense. Can't we just remove the `mpi` variant instead of having a boolean variant and requiring it to be `True`?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/slate/package.py (hunk 1 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/slate/package.py]: I'm following the consensus in the linked issue https://github.com/spack/spack/issues/44952 that it is preferable to avoid breaking current usage where +mpi is explicitly included in the spec.

Comment (alalazo) [reply var/spack/repos/builtin/packages/slate/package.py]: In that case we should add something that says `mpi` is deprecated, and remove the variant at a later point. If we never remove any variant out of concerns for backward compatibility of specs, `package.py` will become full of tech debts. 

Also, we should remove mentions of `+mpi` and `~mpi` in the recipe, since any check done on that variant will lose meaning (it can only be `+mpi`).

Comment (alalazo) [reply var/spack/repos/builtin/packages/slate/package.py]: I think this is related to #38829 I guess my point is, without implementing a deprecation mechanism, like:
```python
variant("foo", ..., deprecated=True)
```
people will not be warned that they need to update a spec. If they're not, we are either delaying the pain of discovering that a variant is gone, or keeping all the variants forever, which in time is not maintainable.

**TL;DR** I'm fine if we want to use `requires` here, but we need to remove conditional checks on `+mpi` from this recipe. I added the feature request above to the v0.23 milestone, so hopefully we'll get a better mechanism in Spack v0.23.

Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/slate/package.py]: I agree with the sentiment about delayed pain.  The message will now advise tha tht evariants are deprecated.

**Review by alalazo — DISMISSED** [review]

**Comment by alalazo** [comment]
Apologies, @G-Ragghianti Can you rebase? I'll merge this asap.

**Comment by G-Ragghianti** [comment]
@alalazo OK it is rebased. Thanks.

**Review by alalazo — APPROVED** [review]
